src/org/opensolaris/opengrok/history/HistoryGuru.java

```diff
@@ -341,69 +343,95 @@ public void addRepositories(File[] files, Collection<RepositoryInfo> repos,
      */
     private void addRepositories(File[] files, Collection<RepositoryInfo> repos,
             IgnoredNames ignoredNames, boolean recursiveSearch, int depth) {
+
+        // This check for ignoring a repository only works for embedded
+        // source code control directories used by SCCS and the like.
         for (File file : files) {
-            Repository repository = null;
-            if (file.getName().equals(".opengrok_skip_history")) {
+            if (file.getName().equals(ignoreRepositoryName)) {
                 LOGGER.log(Level.INFO,
                         "Skipping history cache creation for {0} and its subdirectories",
                         file.getParentFile().getAbsolutePath());
                 return;
             }
         }
         for (File file : files) {
-            Repository repository = null;
+            Boolean ignoreRepository = false;
             try {
-                repository = RepositoryFactory.getRepository(file);
-            } catch (InstantiationException ie) {
-                LOGGER.log(Level.WARNING, "Could not create repository for '"
-                        + file + "', could not instantiate the repository.", ie);
-            } catch (IllegalAccessException iae) {
-                LOGGER.log(Level.WARNING, "Could not create repository for '"
-                        + file + "', missing access rights.", iae);
-            }
-            if (repository == null) {
-                // Not a repository, search its sub-dirs
-                if (file.isDirectory() && !ignoredNames.ignore(file)) {
-                    File subFiles[] = file.listFiles();
-                    if (subFiles == null) {
-                        LOGGER.log(Level.WARNING,
-                                "Failed to get sub directories for ''{0}'', check access permissions.",
-                                file.getAbsolutePath());
-                    } else if (depth <= scanningDepth) {
-                        addRepositories(subFiles, repos, ignoredNames, depth + 1);
-                    }
+                // This check for ignoring a repository is for the
+                // kind of repository, such as AccuRev, where the
+                // entire project folder layout (workspace) is
+                // considered as the repository. Here we need to
+                // look inside the folder for the magic file.
+                if( file.isDirectory() ) {
+                    String path = file.getCanonicalPath();
+                    File skipRepository = new File(path, ignoreRepositoryName);
+                    ignoreRepository = skipRepository.exists();
                 }
+            } catch( IOException exp ) {
+                LOGGER.log(Level.WARNING,
+                    "Failed to get canonical path for {0}: {1}",
+                    new Object[]{file.getAbsolutePath(), exp.getMessage()});
+            }
+            
+            if( ignoreRepository ) {
+                LOGGER.log(Level.INFO,
```

Comment:
rather than introducing another level of indentation, `continue` can be used here